pkg/remotes/docker/artifacts.go

```diff
@@ -0,0 +1,266 @@
+				Committed: true,
+				Status: content.Status{
+					Ref:    d.reference,
+					Total:  desc.Size,
+					Offset: desc.Size,
+				}})
+
+			return errdefs.ErrAlreadyExists
+		} else if resp.StatusCode != http.StatusNotFound {
+			err := remoteserrors.NewUnexpectedStatusErr(resp)
+			resp.Body.Close()
+			return err
+		}
+	}
+
+	return nil
+}
+
+type response struct {
+	*http.Response
+	err error
+}
+
+type artifactsManifest struct {
+	ref       string
+	expected  digest.Digest
+	pipe      *io.PipeWriter
+	tracker   docker.StatusTracker
+	responseC <-chan response
+}
+
+func (pw *artifactsManifest) Write(p []byte) (n int, err error) {
+	status, err := pw.tracker.GetStatus(pw.ref)
+	if err != nil {
+		return n, err
+	}
+	n, err = pw.pipe.Write(p)
+	status.Offset += int64(n)
+	status.UpdatedAt = time.Now()
+	pw.tracker.SetStatus(pw.ref, status)
+	return
+}
+
+func (pw *artifactsManifest) Close() error {
+	return pw.pipe.Close()
+}
+
+func (pw *artifactsManifest) Status() (content.Status, error) {
+	status, err := pw.tracker.GetStatus(pw.ref)
+	if err != nil {
+		return content.Status{}, err
+	}
+	return status.Status, nil
+
+}
+
+func (pw *artifactsManifest) Digest() digest.Digest {
+	return pw.expected
+}
+
+func (pw *artifactsManifest) Commit(ctx context.Context, size int64, expected digest.Digest, opts ...content.Opt) error {
+	// Check whether read has already thrown an error
+	if _, err := pw.pipe.Write([]byte{}); err != nil && err != io.ErrClosedPipe {
+		return errors.Wrap(err, "pipe error before commit")
+	}
+
+	if err := pw.pipe.Close(); err != nil {
+		return err
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case resp := <-pw.responseC:
+		if resp.err != nil {
+			return resp.err
+		}
+		defer resp.Response.Body.Close()
+
+		// 201 is specified return status, some registries return
+		// 200, 202 or 204.
+		switch resp.StatusCode {
+		case http.StatusOK, http.StatusCreated, http.StatusNoContent, http.StatusAccepted:
+		default:
+			return remoteserrors.NewUnexpectedStatusErr(resp.Response)
+		}
+
+		status, err := pw.tracker.GetStatus(pw.ref)
+		if err != nil {
+			return errors.Wrap(err, "failed to get status")
+		}
+
+		if size > 0 && size != status.Offset {
+			return errors.Errorf("unexpected size %d, expected %d", status.Offset, size)
+		}
+
+		if expected == "" {
+			expected = status.Expected
+		}
+
+		actual, err := digest.Parse(resp.Header.Get("Docker-Content-Digest"))
+		if err != nil {
+			return errors.Wrap(err, "invalid content digest in response")
+		}
+
+		if actual != expected {
+			return errors.Errorf("got digest %s, expected %s", actual, expected)
+		}
+
+		status.Committed = true
+		status.UpdatedAt = time.Now()
+		pw.tracker.SetStatus(pw.ref, status)
+	}
+
+	return nil
+}
+
+func (pw *artifactsManifest) Truncate(size int64) error {
+
+	return errors.New("cannot truncate remote upload")
+}
+
+func requestWithMountFrom(req *request, mount, from string) *request {
```

Comment:
where is it used?

Reply:
This is from docker pusher code, but I can remove it